django/db/models/query.py

```diff
@@ -585,21 +585,17 @@ def update_or_create(self, defaults=None, **kwargs):
         defaults = defaults or {}
         self._for_write = True
         with transaction.atomic(using=self.db):
-            try:
-                obj = self.select_for_update().get(**kwargs)
-            except self.model.DoesNotExist:
-                params = self._extract_model_params(defaults, **kwargs)
-                # Lock the row so that a concurrent update is blocked until
-                # after update_or_create() has performed its save.
-                obj, created = self._create_object_from_params(kwargs, params, lock=True)
-                if created:
-                    return obj, created
+            # Lock the row so that a concurrent update is blocked until
+            # after update_or_create() has performed its save.
+            obj, created = self.select_for_update().get_or_create(defaults, **kwargs)
+            if created:
+                return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
             obj.save(using=self.db)
         return obj, False
 
-    def _create_object_from_params(self, lookup, params, lock=False):
+    def _create_object_from_params(self, lookup, params):
```

Comment:
Do we need `_create_object_from_params()` anymore? I think we can move this logic to `get_or_create()` (where it was before adding `update_or_create()`).

Reply:
Done